src/stack.ts

```diff
@@ -255,17 +316,111 @@ export class Stack extends cdk.Stack {
+     * ways that the environment comes into play in a Pulumi CDK application. When resources are created
+     * they are created with a specific provider that is either inherited from the Stack or the App. There
+     * are some values though that CDK generates based on what environment is passed to the StackProps.
+     *
+     * Below is an example of something a user could configure (by mistake).
+     *
+     * @example
+     * new App('testapp', (scope: App) => {
+     *    new Stack(scope, 'teststack', {
+     *        providers: [
+     *          new native.Provider('native-provider', { region: 'us-east-1' }),
+     *        ],
+     *        props: { env: { region: 'us-east-2' }},
+     *    })
+     * }, {
+     *      providers: [
+     *          new native.Provider('native-provider', { region: 'us-west-2' }),
+     *      ]
+     *
+     * })
+     */
+    private validateEnv(): void {
+        const providers = providersToArray(this.options?.providers);
+        const nativeProvider = providers.find((p) => native.Provider.isInstance(p));
+        const awsProvider = providers.find((p) => aws.Provider.isInstance(p));
+
+        const awsRegion = aws.getRegionOutput({}, { parent: this.app, provider: awsProvider }).name;
+        const awsAccount = aws.getCallerIdentityOutput({}, { parent: this.app, provider: awsProvider }).accountId;
+        const nativeRegion = native.getRegionOutput({ parent: this.app, provider: nativeProvider }).region;
+        const nativeAccount = native.getAccountIdOutput({ parent: this.app, provider: nativeProvider }).accountId;
+
+        pulumi
+            .all([awsRegion, awsAccount, nativeRegion, nativeAccount])
+            .apply(([awsRegion, awsAccount, nativeRegion, nativeAccount]) => {
+                // This is to ensure that the user does not pass a different region to the provider and the stack environment.
+                if (!cdk.Token.isUnresolved(this.region) && nativeRegion !== this.region) {
+                    throw new Error(
+                        `The stack '${this.node.id}' has conflicting regions between the native provider (${nativeRegion}) and the stack environment (${this.region}).\n` +
+                            'Please ensure that the stack environment region matches the region of the native provider.',
+                    );
+                }
+
+                if (!cdk.Token.isUnresolved(this.account) && this.account !== nativeAccount) {
+                    throw new Error(
+                        `The stack '${this.node.id}' has conflicting accounts between the native provider (${nativeAccount}) and the stack environment (${this.account}).\n` +
+                            'Please ensure that the stack environment account matches the account of the native provider.',
+                    );
+                }
+
+                if (nativeAccount !== awsAccount) {
+                    warn(
```

Comment:
Using a warning instead of throwing an error because it's possible this could still be correct.

Reply:
In which cases would this be correct? Not for the mapping scenarios, right?
Probably somebody interleaving classic resources with their cdk stack?